src/instructlab/model/train.py

```diff
@@ -521,20 +615,333 @@ def get_files(directory: str, pattern: str) -> list[str]:
+
+    server = None
+    model_serve_url = None
+    try:
+        logger.debug("Starting model server for mt-bench answer generation")
+        server, model_serve_url = launch_server(
+            ctx=ctx,
+            model=str(ckpt_path),
+            model_name=str(ckpt_path),
+            gpus=torch.cuda.device_count(),
+            max_workers=multiprocessing.cpu_count(),
+            enable_serving_output=False,
+            backend=backends.VLLM,
+        )
+        logger.debug("Generating mt-bench answers")
+        evaluator.gen_answers(model_serve_url)
+    finally:
+        if server is not None:
+            server.shutdown()
+
+    try:
+        logger.debug("Starting model server for mt-bench answer judgement")
+        server, model_serve_url = launch_server(
+            ctx=ctx,
+            model=str(mtbench_judge),
+            model_name=str(mtbench_judge),
+            gpus=torch.cuda.device_count(),
+            max_workers=multiprocessing.cpu_count(),
+            backend=backends.VLLM,
+            enable_serving_output=False,
+        )
+        logger.debug("Judging mt-bench answers")
+        mt_bench_results: tuple = evaluator.judge_answers(model_serve_url)
+        ckpt_score: float = mt_bench_results[0]
+    finally:
+        if server is not None:
+            server.shutdown()
+
+    return ckpt_score
+
+
+def _evaluate_dir_of_checkpoints(
+    checkpoints_dir_path: pathlib.Path, eval_func: typing.Callable[..., float]
+) -> tuple[float, pathlib.Path]:
+    """Run eval_func on all model checkpoints in a directory, returning the best performing."""
+
+    # TODO: parallelize MMLU over available GPUs
+
+    results: list[typing.Tuple[float, pathlib.Path]] = []
+    for ckpt_path in checkpoints_dir_path.iterdir():
+        if ckpt_path.is_dir():
+            logger.debug(ckpt_path)
+            ckpt_score = eval_func(ckpt_path=ckpt_path)
             click.secho(
-                f"Error while executing training library {exc}",
+                f"CHECKPOINT EVALUATION: {ckpt_path} SCORED {ckpt_score}",
                 fg="red",
+                bg="cyan",
             )
-            raise click.exceptions.Exit(1)
+            results.append((ckpt_score, ckpt_path))
+
+    if not results:
+        raise RuntimeError(
+            "_evaluate_dir_of_checkpoints - No checkpoints were evaluated successfully at %s. No scores were recored."
+            % checkpoints_dir_path
+        )
+
+    ckpt_performance: list[tuple[float, pathlib.Path]] = sorted(results, reverse=True)
+
+    # (score, path)
+    return ckpt_performance[0]
+
+
+def _run_phased_training(
+    ctx,
+    train_args: TrainingArgs,
+    torch_args: TorchrunArgs,
+    base_dir: pathlib.Path,
+    phase1_data: pathlib.Path,
+    phase1_num_epochs: int | None,
+    phase1_samples_per_save: int | None,
+    phase2_data: pathlib.Path,
+    phase2_num_epochs: int | None,
+    phase2_samples_per_save: int | None,
+    mtbench_judge: pathlib.Path,
```

Comment:
This can be `None` (as per `click` option definition).

Reply:
This *can't* be none, I updated the file so this is checked before entering phased training